applications/AdjointFluidApplication/custom_utilities/numerical_diffusion.h

```diff
@@ -0,0 +1,407 @@
+                            CalculateSVMethodCharacteristicMatrix<TDim+1, TDim>(rElement);
+
+        for (IndexType i=0; i < TDim + 1; i++)
+        {
+            characteristic_matrix(TDim, i) = 0.0;
+            characteristic_matrix(i, TDim) = 0.0;
+        }
+        
+        double velocity_divergence = CalculateVelocityDivergence<TDim>(rElement);
+        characteristic_matrix(TDim, TDim) = 0.5 * velocity_divergence;
+
+        Eigen::JacobiSVD<Eigen::Matrix<double, TDim+1, TDim+1>> svd(
+                characteristic_matrix, Eigen::ComputeThinU | Eigen::ComputeThinV
+                );
+
+        const auto& S = svd.singularValues();
+
+        double numerical_viscosity;
+        numerical_viscosity = S[0];
+
+        return numerical_viscosity;
+    }
+
+    template<unsigned int TDim>
+    double CalculateNumericalDiffusionSVMethodPressureDecoupled(Element& rElement, const ProcessInfo& rCurrentProcessInfo)
+    {
+        Eigen::Matrix<double, TDim, TDim>  characteristic_matrix = 
+                CalculateSVMethodCharacteristicMatrix<TDim, TDim>(rElement);
+
+        Eigen::JacobiSVD<Eigen::Matrix<double, TDim, TDim>> svd(
+            characteristic_matrix, Eigen::ComputeThinU | Eigen::ComputeThinV
+            );
+
+        const auto& S = svd.singularValues();
+
+        double numerical_viscosity;
+        numerical_viscosity = S[0];
+        
+        return numerical_viscosity;    
+    }
+
+    template<unsigned int TDim>
+    double CalculateNumericalDiffusionEigenFullMatrix(Element& rElement, const ProcessInfo& rCurrentProcessInfo)
+    {
+        constexpr unsigned int TNumNodes = TDim + 1;
+
+        BoundedMatrix<double, TNumNodes, TDim> dn_dx;
+        array_1d< double, TNumNodes > n;
+        double volume;
+
+        GeometryUtils::CalculateGeometryData(rElement.GetGeometry(),dn_dx,n,volume);
+
+        Matrix vms_steady_term_primal_gradient;
+
+        rElement.Calculate(
+                            VMS_STEADY_TERM_PRIMAL_GRADIENT_MATRIX,
+                            vms_steady_term_primal_gradient,
+                            rCurrentProcessInfo);
+
+        Vector adjoint_values_vector;
+        rElement.GetValuesVector(adjoint_values_vector, 1);
+
+        BoundedVector<double, (TDim+1)*TDim> temp_1 = prod(vms_steady_term_primal_gradient, adjoint_values_vector);
+        double adjoint_energy = inner_prod(temp_1, adjoint_values_vector);
+        
+        MatrixType numerical_diffusion_matrix;
+        InitializeMatrix<TDim>(numerical_diffusion_matrix);
+
+        AddNumericalDiffusionTerm<TDim>(numerical_diffusion_matrix, dn_dx, volume);
+
+        BoundedVector<double, (TDim+1)*TDim> temp_2 = prod(numerical_diffusion_matrix, adjoint_values_vector);
+        double diffusion_energy = inner_prod(temp_2,adjoint_values_vector);
+
+        double numerical_diffusion = 0.0;
+
+        if (adjoint_energy > 0.0 and diffusion_energy > 0.0)
+            numerical_diffusion = adjoint_energy/diffusion_energy;
```

Comment:
you could add a debug check if diffusion energy is ~0, this would prevent nasty debugging
`KRATOS_DEBUG_ERROR_IF(diffusion_energy < std_num_limits ...)`

Reply:
done